synapse/storage/databases/main/state_deltas.py

```diff
@@ -16,11 +16,17 @@
 from typing import Any, Dict, List, Tuple
 
 from synapse.storage._base import SQLBaseStore
+from synapse.storage.database import LoggingTransaction
+from synapse.util.caches.stream_change_cache import StreamChangeCache
 
 logger = logging.getLogger(__name__)
 
 
 class StateDeltasStore(SQLBaseStore):
+    # This class must be mixed in with a child class which provides the following
+    # attribute. TODO: can we get static analysis to enforce this?
+    _curr_state_delta_stream_cache: StreamChangeCache
```

Comment:
Can we make this an abstract class perhaps, with this being an abstract.. field... if that exists ...

It doesn't seem like it does, but I would be tempted to try an abstract property:

```python
    @property
    @abstractmethod
    def _curr_state_delta_stream_cache(self) -> StreamChangeCache:
        ...
```

Not super clean, but curious if that'd do the trick (and if it's allowed to just overwrite it with a field later)

Reply:
I got

```
synapse/app/generic_worker.py:222: error: Definition of "_curr_state_delta_stream_cache" in base class "StateDeltasStore" is incompatible with definition in base class "SlavedEventStore"  [misc]
```

after

```
class StateDeltasStore(SQLBaseStore, abc.ABC):
    @property
    @abc.abstractmethod
    def _curr_state_delta_stream_cache(self) -> StreamChangeCache:
        """This class must be mixed in with a child class which provides the
        following attribute."""
        ...
```

Reply:
You might be interested in the discussion here: https://github.com/python/mypy/issues/4426